src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py

```diff
@@ -0,0 +1,357 @@
+
+    def __init__(self, *args, **kwargs):
+        super(_ServerSSLCertReloadTest, self).__init__(*args, **kwargs)
+        self.client_key_1_pem = resources.cert_hier_1_client_1_key()
+        self.client_chain_1_pem = '\n'.join([
+            resources.cert_hier_1_client_1_cert(),
+            resources.cert_hier_1_intermediate_ca_cert(),
+        ])
+        self.client_key_2_pem = resources.cert_hier_2_client_1_key()
+        self.client_chain_2_pem = '\n'.join([
+            resources.cert_hier_2_client_1_cert(),
+            resources.cert_hier_2_intermediate_ca_cert(),
+        ])
+        self.server_key_1_pem = resources.cert_hier_1_server_1_key()
+        self.server_cert_1_pem = '\n'.join([
+            resources.cert_hier_1_server_1_cert(),
+            resources.cert_hier_1_intermediate_ca_cert(),
+        ])
+        self.server_key_2_pem = resources.cert_hier_2_server_1_key()
+        self.server_cert_2_pem = '\n'.join([
+            resources.cert_hier_2_server_1_cert(),
+            resources.cert_hier_2_intermediate_ca_cert(),
+        ])
+        self.with_client_auth = True
+        self.server = None
+        self.port = None
+
+    def setUp(self):
+        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+        services_pb2_grpc.add_FirstServiceServicer_to_server(
+            _server_application.FirstServiceServicer(), self.server)
+        switch_cert_on_client_num = 10
+        initial_cert_config = grpc.ssl_server_certificate_config(
+            [(self.server_key_1_pem, self.server_cert_1_pem)],
+            root_certificates=resources.cert_hier_2_root_ca_cert())
+        cert_config_fetcher = CertConfigFetcher(resources.cert_hier_1_root_ca_cert(),
+                                                self.server_key_2_pem,
+                                                self.server_cert_2_pem,
+                                                switch_cert_on_client_num)
+        server_credentials = grpc.ssl_server_credentials_dynamic_cert_config(
+            initial_cert_config,
+            cert_config_fetcher,
+            require_client_auth=self.with_client_auth)
+        self.port = self.server.add_secure_port('[::]:0', server_credentials)
+        self.server.start()
+
+    def tearDown(self):
+        if self.server:
+            self.server.stop(0)
+
+    def _perform_rpc(self, client_stub, expect_success):
+        # we don't care about the actual response of the rpc; only
+        # whether we can perform it or not, and if not, the reason
+        # must be "connect failed"
+        if expect_success:
+            client_stub.UnUn(_application_common.UNARY_UNARY_REQUEST)
+        else:
+            with self.assertRaises(grpc.RpcError) as exception_context:
+                client_stub.UnUn(_application_common.UNARY_UNARY_REQUEST)
+            self.assertEqual(exception_context.exception.code(), grpc.StatusCode.UNAVAILABLE)
+
+    def _create_client_stub(self,
```

Comment:
Insufficient use of `self` parameter to justify placing this at instance-scope. Devolve it to module-scope by swapping `port` for `self`. Since it will be called from the class, place it earlier-in-the-file-than ("above") the class.

Reply:
will do

Reply:
done